README.md

# [MOVED] Analytics JS integration appboy

**This repository has been moved to the open source [Analytics JS Integrations](https://github.com/segmentio/analytics.js-integrations).**

If you want to fork, contribute, or open issues, please do it in the new repository. Existing issues/pull requests will be addressed in the new location.

* [New location](https://github.com/segmentio/analytics.js-integrations/tree/master/integrations/appboy)
* Last release for this repository: [1.7.1](https://github.com/segment-integrations/analytics.js-integration-appboy/releases/tag/1.7.1)
* [Commit](monorepoURL)
* [Previous version of this README](README-OLD.md)

Released under the [MIT license](LICENSE).


HISTORY.md
1.8.0 / 2018-08-02
==================

* Revert "Remove changeUser from track, page, group and completedOrder" (#27).

1.7.0 / 2018-07-31
==================

This release includes two changes:

1. Have Initialization hook into Braze callbacks (#26)

  The `v2` initialization of this integration operates with a synchronous mindset inherited from `v1` of the integration. In fact, the Braze `v2` SDK has callbacks which indicate, among other things, when in-app messaging is available. It is important that this integration not invoke `ready()` until these in-app messages are ready. Otherwise, a race condition results, and users invoking `analytics.page()` are not guaranteed to see the appropriate in-app messages.

2.  Remove changeUser from track, page, group and completedOrder (#27).

  Currently, the Braze `Analytics.js` integration does not properly retrieve the` userId` on `track`, `group`, `page`, and `completedOrder` events. After speaking with the Braze team, they mentioned that we should only be calling `changeUser` on `identify`.

1.6.0 / 2018-07-06
==================

  * Update SDK URLs (#21)

1.5.0 / 2018-07-27
==================

  * Update datacenter mappings to support new Appboy clusters

1.5.0 / 2018-04-11
==================

  * Add support for v2 of Braze SDK

1.4.1 / 2017-08-09
==================

  * Update custom endpoint validation

1.4.0 / 2017-07-19
==================

  * Add ability to add a custom API endpoint

1.3.0 / 2017-07-05
==================

  * Switch the endpoint when datacenter is set to 'eu'

1.2.1 / 2017-06-26
==================

  * Remove reserved keys from custom event properties and user attributes

1.2.0 / 2017-05-15
==================

  * Add enableHtmlInAppMessages option
  * Add support for page calls

1.1.0 / 2017-04-27
==================

  * Add analytics hanlding for client side

1.0.0 / 2017-03-01
==================

  * Initial release

0.0.1 / 2017-02-21
==================

  * Initial scaffold :sparkles:


test/index.test.js
'use strict';

var Analytics = require('@segment/analytics.js-core').constructor;
var integration = require('@segment/analytics.js-integration');
var sandbox = require('@segment/clear-env');
var tester = require('@segment/analytics.js-integration-tester');
var Appboy = require('../lib/');
var assert = require('assert');
var sinon = require('sinon');

describe('Appboy', function() {
  var analytics;
  var appboy;
  var options = {
    apiKey: '7c664901-d8c0-4f82-80bf-e7e7a24478e8',
    automaticallyDisplayMessages: true,
    safariWebsitePushId: '',
    enableHtmlInAppMessages: false,
    trackAllPages: false,
    trackNamedPages: false,
    customEndpoint: '',
    version: 1
  };

  beforeEach(function() {
    analytics = new Analytics();
    appboy = new Appboy(options);
    analytics.use(Appboy);
    analytics.use(tester);
    analytics.add(appboy);
  });

  afterEach(function(done) {
    analytics.restore();
    analytics.reset();
    appboy.reset();
    sandbox();
    done();
  });

  it('should have the right settings', function() {
    analytics.compare(Appboy, integration('Appboy')
      .global('appboy')
      .option('apiKey', '')
      .option('safariWebsitePushId', '')
      .option('allowCrawlerActivity', false)
      .option('doNotLoadFontAwesome', false)
      .option('enableLogging', false)
      .option('automaticallyDisplayMessages', true)
      .option('localization', 'en')
      .option('minimumIntervalBetweenTriggerActionsInSeconds', 30)
      .option('openInAppMessagesInNewTab', false)
      .option('openNewsFeedCardsInNewTab', false)
      .option('sessionTimeoutInSeconds', 30)
      .option('requireExplicitInAppMessageDismissal', false)
      .option('enableHtmlInAppMessages', false)
      .option('trackAllPages', false)
      .option('trackNamedPages', false)
      .option('customEndpoint', '')
      .option('version', 1)
    );
  });

  describe('before loading', function() {
    beforeEach(function() {
      analytics.stub(appboy, 'load');
    });

    describe('#initialize', function() {
      it('should call #load', function() {
        analytics.initialize();
        analytics.called(appboy.load);
      });
    });
  });

  describe('loading', function() {
    it('should load', function(done) {
      analytics.load(appboy, done);
    });

    it('should set the sdk endpoint to the correct datacenter', function(done) {
      var datacenterMappings = {
        us:   'https://sdk.iad-01.braze.com',
        us02: 'https://sdk.iad-02.braze.com',
        us03: 'https://sdk.iad-03.braze.com',
        eu:   'https://sdk.fra-01.braze.eu'
      };

      var spy = sinon.spy(appboy, 'initializeTester');
      for (var option in datacenterMappings) {
        if (!datacenterMappings.hasOwnProperty(option)) continue;
        appboy.options.datacenter = option;
        analytics.once('ready', function() {
          try {
            assert.equal(spy.args[0][1].baseUrl, datacenterMappings[option] + '/api/v3');
            done();
          } catch (e) {
            done(e);
          }
          spy.restore();
        });
        analytics.initialize();
      }
    });

    it('should use initializeV1 if version is set to 1', function(done) {
      var V1spy = sinon.spy(appboy, 'initializeV1');
      var V2spy = sinon.spy(appboy, 'initializeV2');
      appboy.options.version = 1;
      analytics.once('ready', function() {
        try {
          assert(V1spy.called);
          assert(!V2spy.called);
          done();
        } catch (e) {
          done(e);
        }
      });
      analytics.initialize();
    });

    it('should use initializeV2 if version is set to 2', function(done) {
      var V1spy = sinon.spy(appboy, 'initializeV1');
      var V2spy = sinon.spy(appboy, 'initializeV2');
      appboy.options.version = 2;
      analytics.once('ready', function() {
        try {
          assert(V2spy.called);
          assert(!V1spy.called);
          done();
        } catch (e) {
          done(e);
        }
      });
      analytics.initialize();
    });

    it('should set the sdk endpoint to a custom URI if one is provided', function(done) {
      appboy.options.customEndpoint = 'https://my.custom.endpoint.com';
      var spy = sinon.spy(appboy, 'initializeTester');
      analytics.once('ready', function() {
        try {
          assert.equal(spy.args[0][1].baseUrl, 'https://my.custom.endpoint.com/api/v3');
          done();
        } catch (e) {
          done(e);
        }
      });
      analytics.initialize();
    });

    it('should prefix the customEndpoint with https:// if it was not by the user', function(done) {
      appboy.options.customEndpoint = 'my.custom.endpoint.com';
      var spy = sinon.spy(appboy, 'initializeTester');
      analytics.once('ready', function() {
        try {
          assert.equal(spy.args[0][1].baseUrl, 'https://my.custom.endpoint.com/api/v3');
          done();
        } catch (e) {
          done(e);
        }
      });
      analytics.initialize();
    });

    it('should send Safari Website Push ID if provided in the settings', function(done) {
      appboy.options.safariWebsitePushId = 'web.com.example.domain';
      var spy = sinon.spy(appboy, 'initializeTester');
      analytics.once('ready', function() {
        try {
          analytics.assert(spy.args[0][1].safariWebsitePushId, options.safariWebsitePushId);
          done();
        } catch (e) {
          done(e);
        }
      });
      analytics.initialize();
    });

    describe('initializeV1', function() {
      it('should call changeUser if userID is present', function(done) {
        analytics.user().id('user-id');
        analytics.once('ready', function() {
          assert.equal(window.appboy.getUser().getUserId(), 'user-id');
          done();
        });
        analytics.initialize();
      });
    });

    describe('initializeV2', function() {
      it('should call changeUser if userID is present', function(done) {
        appboy.options.version = 2;
        analytics.user().id('user-id');
        analytics.once('ready', function() {
          window.appboy.getUser().getUserId(function(userId) {
            assert.equal(userId, 'user-id');
            done();
          });
        });
        analytics.initialize();
      });

      it('should initialize the appboy sdk with default options if none are provided', function(done) {
        appboy.options.version = 2;
        var spy = sinon.spy(appboy, 'initializeTester');
        var defaults = {
          allowCrawlerActivity: false,
          doNotLoadFontAwesome: false,
          enableLogging: false,
          safariWebsitePushId: '',
          localization: 'en',
          minimumIntervalBetweenTriggerActionsInSeconds: 30,
          openInAppMessagesInNewTab: false,
          openNewsFeedCardsInNewTab: false,
          sessionTimeoutInSeconds: 30,
          requireExplicitInAppMessageDismissal: false,
          enableHtmlInAppMessages: false
        };
        analytics.once('ready', function() {
          try {
            assert.deepEqual(spy.args[0][1], defaults);
            done();
          } catch (e) {
            done(e);
          }
        });
        analytics.initialize();
      });
    });
  });

  describe('after loading', function() {
    beforeEach(function(done) {
      analytics.once('ready', done);
      analytics.initialize();
    });

    describe('#identify', function() {
      beforeEach(function() {
        analytics.stub(window.appboy, 'changeUser');
        analytics.stub(window.appboy.ab.User.prototype, 'setFirstName');
        analytics.stub(window.appboy.ab.User.prototype, 'setLastName');
        analytics.stub(window.appboy.ab.User.prototype, 'setPhoneNumber');
        analytics.stub(window.appboy.ab.User.prototype, 'setEmail');
        analytics.stub(window.appboy.ab.User.prototype, 'setAvatarImageUrl');
        analytics.stub(window.appboy.ab.User.prototype, 'setHomeCity');
        analytics.stub(window.appboy.ab.User.prototype, 'setCountry');
        analytics.stub(window.appboy.ab.User.prototype, 'setDateOfBirth');
        analytics.stub(window.appboy.ab.User.prototype, 'setGender');
        analytics.stub(window.appboy.ab.User.prototype, 'setCustomUserAttribute');
      });

      it('should call each Appboy method for standard traits', function() {
        analytics.identify('userId', {
          firstName: 'Alex',
          lastName: 'Noonan',
          phone: '555-555-5555',
          email: 'alex@email.com',
          avatar: 'https://s-media-cache-ak0.pinimg.com/736x/39/b9/75/39b9757ac27c6eabba292d71a63def2c.jpg',
          gender: 'woman',
          birthday: '1991-09-16T00:00:00.000Z',
          address: {
            city: 'Dublin',
            country: 'Ireland'
          }
        });
        analytics.called(window.appboy.changeUser, 'userId');
        analytics.called(window.appboy.ab.User.prototype.setAvatarImageUrl, 'https://s-media-cache-ak0.pinimg.com/736x/39/b9/75/39b9757ac27c6eabba292d71a63def2c.jpg');
        analytics.called(window.appboy.ab.User.prototype.setCountry, 'Ireland');
        analytics.called(window.appboy.ab.User.prototype.setDateOfBirth, 1991, 9, 16);
        analytics.called(window.appboy.ab.User.prototype.setEmail, 'alex@email.com');
        analytics.called(window.appboy.ab.User.prototype.setFirstName, 'Alex');
        analytics.called(window.appboy.ab.User.prototype.setHomeCity, 'Dublin');
        analytics.called(window.appboy.ab.User.prototype.setGender, window.appboy.ab.User.Genders.FEMALE);
        analytics.called(window.appboy.ab.User.prototype.setLastName, 'Noonan');
        analytics.called(window.appboy.ab.User.prototype.setPhoneNumber, '555-555-5555');
      });

      it('should set gender to male when passed male gender', function() {
        analytics.identify('userId', {
          gender: 'male'
        });
        analytics.called(window.appboy.changeUser, 'userId');
        analytics.called(window.appboy.ab.User.prototype.setGender, window.appboy.ab.User.Genders.MALE);
      });

      it('should set gender to other when passed other gender', function() {
        analytics.identify('userId', {
          gender: 'o'
        });
        analytics.called(window.appboy.changeUser, 'userId');
        analytics.called(window.appboy.ab.User.prototype.setGender, window.appboy.ab.User.Genders.OTHER);
      });

      it('should handle custom traits of all types', function() {
        analytics.identify('userId', {
          song: 'Who\'s That Chick?',
          artists: ['David Guetta', 'Rihanna'],
          number: 16,
          date: 'Tue Apr 25 2017 14:22:48 GMT-0700 (PDT)'
        });
        analytics.called(window.appboy.changeUser, 'userId');
        analytics.called(window.appboy.ab.User.prototype.setCustomUserAttribute, 'song', 'Who\'s That Chick?');
        analytics.called(window.appboy.ab.User.prototype.setCustomUserAttribute, 'artists', ['David Guetta', 'Rihanna']);
        analytics.called(window.appboy.ab.User.prototype.setCustomUserAttribute, 'number', 16);
        analytics.called(window.appboy.ab.User.prototype.setCustomUserAttribute, 'date', 'Tue Apr 25 2017 14:22:48 GMT-0700 (PDT)');
      });

      it('should not let you set reserved keys as custom attributes', function() {
        analytics.identify('rick sanchez', { avatar: 'airbender', external_id: '123' });
        analytics.didNotCall(window.appboy.ab.User.prototype.setCustomUserAttribute);
      });
    });

    describe('#group', function() {
      beforeEach(function() {
        analytics.stub(window.appboy.ab.User.prototype, 'setCustomUserAttribute');
      });

      it('should send group calls with group ID as a custom field', function() {
        analytics.group('0e8c78ea9d97a7b8185e8632', {
          name: 'Initech'
        });
        analytics.called(window.appboy.ab.User.prototype.setCustomUserAttribute, 'ab_segment_group_0e8c78ea9d97a7b8185e8632', true);
      });
    });

    describe('#track', function() {
      beforeEach(function() {
        analytics.stub(window.appboy, 'logCustomEvent');
        analytics.stub(window.appboy, 'logPurchase');
      });

      it('should send an event', function() {
        analytics.track('event');

        appboy._onMessagingReady(function() {
          analytics.called(window.appboy.logCustomEvent, 'event');
        });
      });

      it('should send all properties', function() {
        analytics.track('event with properties', {
          nickname: 'noonz',
          spiritAnimal: 'rihanna',
          best_friend: 'han'
        });

        appboy._onMessagingReady(function() {
          analytics.called(window.appboy.logCustomEvent, 'event with properties', {
            nickname: 'noonz',
            spiritAnimal: 'rihanna',
            best_friend: 'han'
          });
        });
      });

      it('should remove reserved keys from custom event properties', function() {
        analytics.track('event with properties', {
          product_id: 'noonz',
          event_name: 'rihanna',
          time: 'han',
          currency: 'usd',
          quantity: '123'
        });

        appboy._onMessagingReady(function() {
          analytics.called(window.appboy.logCustomEvent, 'event with properties', {});
        });
      });

      it('should call logPurchase for each product in a Completed Order event', function() {
        analytics.track('Order Completed', {
          total: 30,
          revenue: 25,
          shipping: 3,
          currency: 'USD',
          products: [
            {
              product_id: '507f1f77bcf86cd799439011',
              sku: '45790-32',
              name: 'Monopoly: 3rd Edition',
              price: 19.23,
              quantity: 1,
              category: 'Games'
            },
            {
              product_id: '505bd76785ebb509fc183733',
              sku: '46493-32',
              name: 'Uno Card Game',
              price: 3,
              quantity: 2,
              category: 'Games'
            }
          ]
        });

        appboy._onMessagingReady(function() {
          analytics.called(window.appboy.logPurchase, '507f1f77bcf86cd799439011', 19.23, 'USD', 1, {
            total: 30,
            revenue: 25,
            shipping: 3
          });
          analytics.called(window.appboy.logPurchase, '505bd76785ebb509fc183733', 3, 'USD', 2, {
            total: 30,
            revenue: 25,
            shipping: 3
          });
        });
      });

      it('should not fail if currency, quantity, and purchaseProperties are undefined', function() {
        analytics.track('Order Completed', {
          products: [
            {
              product_id: '507f1f77bcf86cd799439011',
              price: 17.38
            },
            {
              product_id: '505bd76785ebb509fc183733',
              price: 3
            }
          ]
        });

        appboy._onMessagingReady(function() {
          analytics.called(window.appboy.logPurchase, '507f1f77bcf86cd799439011', 17.38);
          analytics.called(window.appboy.logPurchase, '505bd76785ebb509fc183733', 3);
        });
      });
    });

    describe('#page', function() {
      beforeEach(function() {
        analytics.stub(window.appboy, 'logCustomEvent');
      });

      it('should send a page view if trackAllPages is enabled', function() {
        appboy.options.trackAllPages = true;
        analytics.page();

        appboy._onMessagingReady(function() {
          analytics.called(window.appboy.logCustomEvent, 'Loaded a Page');
        });
      });

      it('should send a page view if trackNamedPages is enabled', function() {
        appboy.options.trackNamedPages = true;
        analytics.page('Home');

        appboy._onMessagingReady(function() {
          analytics.called(window.appboy.logCustomEvent, 'Viewed Home Page');
        });
      });

      it('should not send a page view if trackAllPages and trackNamedPages are disabled', function() {
        analytics.page('Home');

        appboy._onMessagingReady(function() {
          analytics.didNotCall(window.appboy.logCustomEvent);
        });
      });

      it('should not send a page view if trackNamedPages is enabled and name is null', function() {
        analytics.page();

        appboy._onMessagingReady(function() {
          analytics.didNotCall(window.appboy.logCustomEvent);
        });
      });

      it('should send all properties', function() {
        appboy.options.trackAllPages = true;
        analytics.page('Home', {
          title: 'noonz',
          url: 'www.google.com'
        });

        appboy._onMessagingReady(function() {
          analytics.called(window.appboy.logCustomEvent, 'Viewed Home Page', {
            title: 'noonz',
            url: 'www.google.com',
            name: 'Home',
            path: window.location.pathname,
            referrer: window.document.referrer,
            search: ''
          });
        });
      });
    });
  });
});


lib/index.js
'use strict';

/**
 * Module dependencies.
 */

var integration = require('@segment/analytics.js-integration');
var Track = require('segmentio-facade').Track;
var each = require('@ndhoule/each');
var del = require('obj-case').del;
var clone = require('@ndhoule/clone');

/**
 * Expose `Appboy` integration.
 */

var Appboy = module.exports = integration('Appboy')
  .global('appboy')
  .option('apiKey', '')
  .option('safariWebsitePushId', '')
  .option('allowCrawlerActivity', false)
  .option('doNotLoadFontAwesome', false)
  .option('enableLogging', false)
  .option('automaticallyDisplayMessages', true)
  .option('localization', 'en')
  .option('minimumIntervalBetweenTriggerActionsInSeconds', 30)
  .option('openInAppMessagesInNewTab', false)
  .option('openNewsFeedCardsInNewTab', false)
  .option('sessionTimeoutInSeconds', 30)
  .option('requireExplicitInAppMessageDismissal', false)
  .option('enableHtmlInAppMessages', false)
  .option('trackAllPages', false)
  .option('trackNamedPages', false)
  .option('customEndpoint', '')
  .option('version', 1)
  .tag('v1', '<script src="https://js.appboycdn.com/web-sdk/1.6/appboy.min.js">')
  .tag('v2', '<script src="https://js.appboycdn.com/web-sdk/2.1/appboy.min.js">');


Appboy.prototype.initialize = function() {
  var options = this.options;
  var customEndpoint;
  // Setup custom endpoints
  if (options.customEndpoint) {
    var endpoint = options.customEndpoint;
    var regex = new RegExp('^(http|https)://', 'i');
    customEndpoint = (regex.test(endpoint) ? endpoint : 'https://' + endpoint) + '/api/v3';
  } else if (options.datacenter === 'eu') {
    customEndpoint = 'https://sdk.fra-01.braze.eu/api/v3';
  }
  
  if (Number(options.version) === 2) {
    this.initializeV2(customEndpoint);
  } else {
    this.initializeV1(customEndpoint);
  }
};
/**
 * Initialize v1.
 *
 * @api public
 */

Appboy.prototype.initializeV1 = function(customEndpoint) {
  var options = this.options;
  var self = this;
  var userId = this.analytics.user().id();

  // stub out function
  /* eslint-disable */
  +function(a,p,P,b,y) {
    window.appboy={};for(var s="destroy toggleAppboyLogging setLogger openSession changeUser requestImmediateDataFlush requestFeedRefresh subscribeToFeedUpdates logCardImpressions logCardClick logFeedDisplayed requestInAppMessageRefresh logInAppMessageImpression logInAppMessageClick logInAppMessageButtonClick subscribeToNewInAppMessages removeSubscription removeAllSubscriptions logCustomEvent logPurchase isPushSupported isPushBlocked isPushGranted isPushPermissionGranted registerAppboyPushMessages unregisterAppboyPushMessages submitFeedback ab ab.User ab.User.Genders ab.User.NotificationSubscriptionTypes ab.User.prototype.getUserId ab.User.prototype.setFirstName ab.User.prototype.setLastName ab.User.prototype.setEmail ab.User.prototype.setGender ab.User.prototype.setDateOfBirth ab.User.prototype.setCountry ab.User.prototype.setHomeCity ab.User.prototype.setEmailNotificationSubscriptionType ab.User.prototype.setPushNotificationSubscriptionType ab.User.prototype.setPhoneNumber ab.User.prototype.setAvatarImageUrl ab.User.prototype.setLastKnownLocation ab.User.prototype.setUserAttribute ab.User.prototype.setCustomUserAttribute ab.User.prototype.addToCustomAttributeArray ab.User.prototype.removeFromCustomAttributeArray ab.User.prototype.incrementCustomUserAttribute ab.InAppMessage ab.InAppMessage.SlideFrom ab.InAppMessage.ClickAction ab.InAppMessage.DismissType ab.InAppMessage.OpenTarget ab.InAppMessage.ImageStyle ab.InAppMessage.Orientation ab.InAppMessage.CropType ab.InAppMessage.prototype.subscribeToClickedEvent ab.InAppMessage.prototype.subscribeToDismissedEvent ab.InAppMessage.prototype.removeSubscription ab.InAppMessage.prototype.removeAllSubscriptions ab.InAppMessage.Button ab.InAppMessage.Button.prototype.subscribeToClickedEvent ab.InAppMessage.Button.prototype.removeSubscription ab.InAppMessage.Button.prototype.removeAllSubscriptions ab.SlideUpMessage ab.ModalMessage ab.FullScreenMessage ab.ControlMessage ab.Feed ab.Feed.prototype.getUnreadCardCount ab.Card ab.ClassicCard ab.CaptionedImage ab.Banner ab.WindowUtils display display.automaticallyShowNewInAppMessages display.showInAppMessage display.showFeed display.destroyFeed display.toggleFeed sharedLib".split(" "),i=0;i<s.length;i++){for(var k=appboy,l=s[i].split("."),j=0;j<l.length-1;j++)k=k[l[j]];k[l[j]]=function(){console&&console.error("The Appboy SDK has not yet been loaded.")}}appboy.initialize=function(){console&&console.error("Appboy cannot be loaded - this is usually due to strict corporate firewalls or ad blockers.")};appboy.getUser=function(){return new appboy.ab.User};appboy.getCachedFeed=function(){return new appboy.ab.Feed};
  }(document, 'script', 'link');
  /* eslint-enable */
  
  // this is used to test this.loaded
  this._shim = window.appboy.initialize;

  this.load('v1', function() {
    var config = {};
    var datacenterMappings = {
      us:   'https://sdk.iad-01.braze.com',
      us02: 'https://sdk.iad-02.braze.com',
      us03: 'https://sdk.iad-03.braze.com',
      eu:   'https://sdk.fra-01.braze.eu'
    };
    if (options.safariWebsitePushId) config.safariWebsitePushId = options.safariWebsitePushId;
    if (options.enableHtmlInAppMessages) config.enableHtmlInAppMessages = true;

    // Setup custom endpoints
    if (options.customEndpoint) {
      var endpoint = options.customEndpoint;
      var regex = new RegExp('^(http|https)://', 'i');
      config.baseUrl = (regex.test(endpoint) ? endpoint : 'https://' + endpoint) + '/api/v3';
    } else {
      config.baseUrl = (datacenterMappings[options.datacenter] || 'https://sdk.iad-01.braze.com') + '/api/v3';
    }

    if (customEndpoint) config.baseUrl = customEndpoint;

    self.initializeTester(options.apiKey, config);
    window.appboy.initialize(options.apiKey, config);

    if (options.automaticallyDisplayMessages) window.appboy.display.automaticallyShowNewInAppMessages();
    if (userId) window.appboy.changeUser(userId);

    window.appboy.openSession();
    self.ready();
  });
};
  
/**
 * Initialize v2.
 *
 * @api public
 */

Appboy.prototype.initializeV2 = function(customEndpoint) {
  var options = this.options;
  var userId = this.analytics.user().id();
  
  /* eslint-disable */
  +function (a, p, P, b, y) {
    window.appboy = {}; window.appboyQueue = [];
    for (var s = "initialize destroy getDeviceId toggleAppboyLogging setLogger openSession changeUser requestImmediateDataFlush requestFeedRefresh subscribeToFeedUpdates logCardImpressions logCardClick logFeedDisplayed requestInAppMessageRefresh logInAppMessageImpression logInAppMessageClick logInAppMessageButtonClick logInAppMessageHtmlClick subscribeToNewInAppMessages removeSubscription removeAllSubscriptions logCustomEvent logPurchase isPushSupported isPushBlocked isPushGranted isPushPermissionGranted registerAppboyPushMessages unregisterAppboyPushMessages submitFeedback trackLocation stopWebTracking resumeWebTracking wipeData ab ab.User ab.User.Genders ab.User.NotificationSubscriptionTypes ab.User.prototype.getUserId ab.User.prototype.setFirstName ab.User.prototype.setLastName ab.User.prototype.setEmail ab.User.prototype.setGender ab.User.prototype.setDateOfBirth ab.User.prototype.setCountry ab.User.prototype.setHomeCity ab.User.prototype.setLanguage ab.User.prototype.setEmailNotificationSubscriptionType ab.User.prototype.setPushNotificationSubscriptionType ab.User.prototype.setPhoneNumber ab.User.prototype.setAvatarImageUrl ab.User.prototype.setLastKnownLocation ab.User.prototype.setUserAttribute ab.User.prototype.setCustomUserAttribute ab.User.prototype.addToCustomAttributeArray ab.User.prototype.removeFromCustomAttributeArray ab.User.prototype.incrementCustomUserAttribute ab.User.prototype.addAlias ab.InAppMessage ab.InAppMessage.SlideFrom ab.InAppMessage.ClickAction ab.InAppMessage.DismissType ab.InAppMessage.OpenTarget ab.InAppMessage.ImageStyle ab.InAppMessage.TextAlignment ab.InAppMessage.Orientation ab.InAppMessage.CropType ab.InAppMessage.prototype.subscribeToClickedEvent ab.InAppMessage.prototype.subscribeToDismissedEvent ab.InAppMessage.prototype.removeSubscription ab.InAppMessage.prototype.removeAllSubscriptions ab.InAppMessage.Button ab.InAppMessage.Button.prototype.subscribeToClickedEvent ab.InAppMessage.Button.prototype.removeSubscription ab.InAppMessage.Button.prototype.removeAllSubscriptions ab.SlideUpMessage ab.ModalMessage ab.FullScreenMessage ab.HtmlMessage ab.ControlMessage ab.Feed ab.Feed.prototype.getUnreadCardCount ab.Card ab.ClassicCard ab.CaptionedImage ab.Banner ab.WindowUtils display display.automaticallyShowNewInAppMessages display.showInAppMessage display.showFeed display.destroyFeed display.toggleFeed sharedLib".split(" "), i = 0; i < s.length; i++){
      for (var m = s[i], k = appboy, l = m.split("."), j = 0; j < l.length - 1; j++)k = k[l[j]]; k[l[j]] = (new Function("return function " + m.replace(/\./g, "_") + "(){appboyQueue.push(arguments); return true}"))()
    }
    appboy.getUser = function () { return new appboy.ab.User };
    appboy.getCachedFeed = function () { return new appboy.ab.Feed };
  }(window, document, 'script');
  /* eslint-enable */

  // https://js.appboycdn.com/web-sdk/2.0/doc/module-appboy.html#.initialize
  var config = {
    safariWebsitePushId: options.safariWebsitePushId,
    enableHtmlInAppMessages: options.enableHtmlInAppMessages,
    allowCrawlerActivity: options.allowCrawlerActivity,
    doNotLoadFontAwesome: options.doNotLoadFontAwesome,
    enableLogging: options.enableLogging,
    localization: options.localization,
    minimumIntervalBetweenTriggerActionsInSeconds: Number(options.minimumIntervalBetweenTriggerActionsInSeconds) || 30,
    openInAppMessagesInNewTab: options.openInAppMessagesInNewTab,
    openNewsFeedCardsInNewTab: options.openNewsFeedCardsInNewTab,
    requireExplicitInAppMessageDismissal: options.requireExplicitInAppMessageDismissal,
    sessionTimeoutInSeconds: Number(options.sessionTimeoutInSeconds) || 30
  };

  if (customEndpoint) config.baseUrl = customEndpoint;
  
  this.initializeTester(options.apiKey, config);
  window.appboy.initialize(options.apiKey, config);
  
  if (options.automaticallyDisplayMessages) window.appboy.display.automaticallyShowNewInAppMessages();
  if (userId) window.appboy.changeUser(userId);
  
  window.appboy.openSession();

  this.load('v2', this.ready);
};

// This is used to test window.appboy.initialize
Appboy.prototype.initializeTester = function() {};

/**
 * Loaded?
 *
 * @api private
 * @return {boolean}
 */

Appboy.prototype.loaded = function() {
  var options = this.options;
  if (Number(options.version) === 2) return window.appboyQueue === null;
  return window.appboy && window.appboy.initialize !== this._shim;
};

/**
 * Identify.
 *
 * @api public
 * @param {Identify} identify
 */

Appboy.prototype.identify = function(identify) {
  var userId = identify.userId();
  var address = identify.address();
  var avatar = identify.avatar();
  var birthday = identify.birthday();
  var email = identify.email();
  var firstName = identify.firstName();
  var gender = identify.gender();
  var lastName = identify.lastName();
  var phone = identify.phone();
  var traits = clone(identify.traits());

  window.appboy.changeUser(userId);
  window.appboy.getUser().setAvatarImageUrl(avatar);
  window.appboy.getUser().setEmail(email);
  window.appboy.getUser().setFirstName(firstName);
  window.appboy.getUser().setGender(getGender(gender));
  window.appboy.getUser().setLastName(lastName);
  window.appboy.getUser().setPhoneNumber(phone);
  if (address) {
    window.appboy.getUser().setCountry(address.country);
    window.appboy.getUser().setHomeCity(address.city);
  }
  if (birthday) {
    window.appboy.getUser().setDateOfBirth(birthday.getUTCFullYear(), birthday.getUTCMonth() + 1, birthday.getUTCDate());
  }

  // delete all the standard traits from traits clone so that we can use appboy's setCustomAttribute on non-standard traits
  // also remove all reserved keys so we dont set them as custom attributes, otherwise Appboy rejects the entire event
  // https://www.appboy.com/documentation/Platform_Wide/#reserved-keys
  var reserved = ['avatar', 'address', 'birthday', 'email', 'id', 'firstName', 'gender', 'lastName', 'phone', 'facebook', 'twitter', 'first_name', 'last_name', 'dob', 'external_id', 'country', 'home_city', 'bio', 'gender', 'phone', 'email_subscribe', 'push_subscribe'];
  each(function(key) {
    delete traits[key];
  }, reserved);

  each(function(value, key) {
    window.appboy.getUser().setCustomUserAttribute(key, value);
  }, traits);
};

/**
 * Group.
 *
 * Sets the group Id for users.
 *
 * @api public
 * @param {Group} group
 */

Appboy.prototype.group = function(group) {
  var userId = group.userId();
  var groupIdKey = 'ab_segment_group_' + group.groupId();

  window.appboy.changeUser(userId);
  window.appboy.getUser().setCustomUserAttribute(groupIdKey, true);
};

/**
 * Track.
 *
 * https://js.appboycdn.com/web-sdk/latest/doc/module-appboy.html#.logCustomEvent
 *
 * @api public
 * @param {Track} track
 */

Appboy.prototype.track = function(track) {
  var userId = track.userId();
  var eventName = track.event();
  var properties = track.properties();
  // remove reserved keys from custom event properties
  // https://www.appboy.com/documentation/Platform_Wide/#reserved-keys
  var reserved = ['time', 'product_id', 'quantity', 'event_name', 'price', 'currency'];
  each(function(key) {
    delete properties[key];
  }, reserved);

  // For temporary testing purposes, attempt to detect errors in calls to
  // `changeUser`.
  try {
    window.appboy.changeUser(userId);
  } catch (e) {
    console.error(e);
  }

  window.appboy.logCustomEvent(eventName, properties);
};

/**
 * Page.
 *
 * https://js.appboycdn.com/web-sdk/latest/doc/module-appboy.html#.logCustomEvent
 *
 * @api public
 * @param {Page} page
 */

Appboy.prototype.page = function(page) {
  var settings = this.options;
  if (!settings.trackAllPages && !settings.trackNamedPages) return;
  if (settings.trackNamedPages && !page.name()) return;

  var userId = page.userId();
  var pageEvent = page.track(page.fullName());
  var eventName = pageEvent.event();
  var properties = page.properties();

  window.appboy.changeUser(userId);
  window.appboy.logCustomEvent(eventName, properties);
};

/**
 * Order Completed.
 *
 * Breaking this out because it requires certain properties that all other events don't.
 *
 * https://js.appboycdn.com/web-sdk/latest/doc/module-appboy.html#.logPurchase
 *
 * @api public
 * @param {Track} track
 */


Appboy.prototype.orderCompleted = function(track) {
  var userId = track.userId();
  var products = track.products();
  var currencyCode = track.currency();
  var purchaseProperties = track.properties();

  window.appboy.changeUser(userId);

  // remove reduntant properties
  del(purchaseProperties, 'products');
  del(purchaseProperties, 'currency');

  // we have to make a separate call to appboy for each product
  each(function(product) {
    var track = new Track({ properties: product });
    var productId = track.productId();
    var price = track.price();
    var quantity = track.quantity();
    window.appboy.logPurchase(productId, price, currencyCode, quantity, purchaseProperties);
  }, products);
};

/**
 * Get Gender.
 *
 * Returns Gender in the way that Appboy understands it.
 *
 * https://js.appboycdn.com/web-sdk/latest/doc/ab.User.html#toc4
 *
 * @api public
 * @param {string} gender
 * @return {string}
 */

function getGender(gender) {
  if (!gender) return;
  if (typeof gender !== 'string') return;

  var femaleGenders = ['woman', 'female', 'w', 'f'];
  var maleGenders = ['man', 'male', 'm'];
  var otherGenders = ['other', 'o'];

  if (femaleGenders.indexOf(gender.toLowerCase()) > -1) return window.appboy.ab.User.Genders.FEMALE;
  if (maleGenders.indexOf(gender.toLowerCase()) > -1) return window.appboy.ab.User.Genders.MALE;
  if (otherGenders.indexOf(gender.toLowerCase()) > -1) return window.appboy.ab.User.Genders.OTHER;
}
